solutions/usaco-815.mdx

```diff
@@ -0,0 +1,115 @@
+---
+id: usaco-815
+source: USACO Gold 2018 Febuary
+title: Taming the Herd
+author: Shijie Ren
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_taming_gold_feb18.html)
+
+## Explanation
+
+Let $\texttt{dp}[i][j]$ be the the minimum number of changes that must be made
+to the first $j$ entries so that there are $i$ breakouts among the first $j$
+entries.
+
+There are three cases when we calculate the current value of $\texttt{dp}[i][j]$:
+
+- If the log was tampered with and the cows would NOT breakout:
+
+  - Compared to the previous day, the breakout number won't change. However, we need one
+    more change of log, so we transfer from $\texttt{dp}[i][j-1]$.
+    $$
+    \texttt{dp}[i][j]=\texttt{dp}[i][j-1]+1
+    $$
+
+- If log was tampered with and cows would breakout (Notice $a[j] \neq 0$):
+
+  - Compared to the previous day the breakout number increases by 1. We also
+    need one more change of the log, so we transfer from $\texttt{dp}[i-1][j-1]$.
+    $$
+    \texttt{dp}[i][j]=\texttt{dp}[i-1][j-1]+1
+    $$
```

Comment:
@dongliu0426 @maggie-j-liu can one of you guys explain this to me? i'm being dumb here lol

Reply:
the dp state is `dp[i][j]` (`i` breakouts in the first `j` entries of the log).

this case is assuming that the entry originally said that the cows won't break out (`a[i] > 0`) but they actually did but just changed the log. the value is `dp[i - 1][j - 1] + 1` because you're adding one more breakout (`i - 1`) from the first `j - 1` entries and one change (`+ 1`)

Reply:
but that's assuming the log would need to be changed, no?

Reply:
yes

Reply:
but that wouldn't always be needed?

Reply:
so what should I do here to (maybe clarify) it?

Reply:
i mean can you explain to me why a change is always necessitated? then i can probably tell you how to reword it

Reply:
> if the log was NOT tampered

In this case,  the i-th entry didn't need a change.
And since the i-th entry was correct, previous entries need to be corrected.

Reply:
but what if a[j] is already the desired value?

Reply:
previous entries (j-a[j] -> j-1) need to be corrected if they didn't match a[j]

Reply:
that...doesn't answer my question?

Reply:
desired value means that this entry was correct , right

Reply:
yeah?

Reply:
so the true entries of the i-th breakout could be determined, transfer from the last state with (i-1)-th breakout?